### ISSUE TITLE
Include AdamW

### ISSUE THREAD

**Issue by GilesStrong** [GilesStrong/lumin#62]
AdamW ([Loshchilov, Hutter, 2017](https://arxiv.org/abs/1711.05101)) should really be included for cases when the user wishes to use weight decay (PyTorch Adam says 'weight_decay', but implements L_2). PyTorch >= 1.2 includes AdamW, and there is also an Apache2.0 implementation [here](https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py). AdamW should be added as an easily accessible optimiser in LUMIN (if not the default). Warnings should also be addded to recommend that the user use AdamW if they specify a weight decay and Adam. The user must actively make this change; LUMIN should not silently switch the optimiser to AdamW.

## Probably solution

LUMIN moves to the latest version of PyTorch, which includes AdamW, but checks will need to be made to make sure there aren't any problems in compatibility.

**Comment by kiryteo** [comment]
As [AdamW](https://pytorch.org/docs/stable/optim.html) is available in torch.optim (as mentioned in issue description), does the modification expect something more than adding 'adamW' after [this](https://github.com/GilesStrong/lumin/blob/master/lumin/nn/models/model_builder.py#L216) as another optimizer choice?

**Comment by GilesStrong** [comment]
Yes, adding `'adamw'` to that part of the code would be enough to get it used by passing a string argument in the `opt_args`, however it might cause an error if a user tries to use it with a lower version of PyTorch (current requirement for lumin is torch > 1.0.0).

Either we could increase the minimum version to 1.2.0 (I'm currently running 1.5.1 and Lumin works fine), or there could be a check on the torch version to make sure that AdamW is present.

There should also be the recommendation to the user to use AdamW if they request Adam and L2 regularisation. This should probably go prior to [this line](https://github.com/GilesStrong/lumin/blob/f89c4250a72059639c8b9ac9c53f8c8a06ff87dc/lumin/nn/models/model_builder.py#L193)
Additionally, [these lines](https://github.com/GilesStrong/lumin/blob/f89c4250a72059639c8b9ac9c53f8c8a06ff87dc/lumin/nn/models/model_builder.py#L190-L191) set the default optimiser to Adam, if no optimiser was specified. This could be improved to change the default optimiser to AdamW if no optimisre was specified but L2 regularisation was specified, and the torch version is high enough.

**Comment by kiryteo** [comment]
Alright, thanks for the pointers! I can add the changes along with the assertions/ warnings for the user. This could be a starting step, later we can just modify the requirements for PyTorch version >= 1.2.0 and mention in the readme as well.

**Comment by GilesStrong** [comment]
That would be great, thank you!

**Comment by kiryteo** [comment]
@GilesStrong The updated model_builder module has AdamW included but other checks as discussed above are currently commented out. In case we use 'regularizer' as an argument, [this](https://github.com/GilesStrong/lumin/blob/master/lumin/nn/models/model_builder.py#L221) will have to be changed accordingly.
